libraries/eosiolib/contracts/eosio/key_value.hpp

```diff
@@ -0,0 +1,1005 @@
+
+               eosio::check(copy_size == tbl_key.size(), "Attempted to update an existing secondary index.");
+               auto res = memcmp(buffer, tbl_key.data(), copy_size);
+               eosio::check(res == 0, "Attempted to update an existing secondary index.");
+
+               if (copy_size > detail::max_stack_buffer_size) {
+                  free(buffer);
+               }
+            } else {
+               auto old_sec_key = table_key(make_prefix(table_name, idx->index_name), idx->get_key(old_value));
+               internal_use_do_not_use::kv_erase(db_name, contract_name.value, old_sec_key.data(), old_sec_key.size());
+               internal_use_do_not_use::kv_set(db_name, contract_name.value, sec_tbl_key.data(), sec_tbl_key.size(), tbl_key.data(), tbl_key.size());
+            }
+         }
+
+      }
+
+      size_t data_size = get_size(value);
+      void* data_buffer = data_size > detail::max_stack_buffer_size ? malloc(data_size) : alloca(data_size);
+
+      serialize(value, data_buffer, data_size);
+
+      internal_use_do_not_use::kv_set(db_name, contract_name.value, tbl_key.data(), tbl_key.size(), (const char*)data_buffer, data_size);
+
+      if (data_size > detail::max_stack_buffer_size) {
+         free(data_buffer);
+      }
+   }
+
+   /**
+    * Removes a value from the table.
+    * @ingroup keyvalue
+    *
+    * @param key - The key of the value to be removed.
+    */
+   void erase(const T& value) {
+      uint32_t value_size;
+
+      auto primary_key = primary_index->get_key(value);
+      auto tbl_key = table_key(make_prefix(table_name, primary_index->index_name), primary_key);
+      auto primary_key_found = internal_use_do_not_use::kv_get(db_name, contract_name.value, tbl_key.data(), tbl_key.size(), value_size);
+
+      if (!primary_key_found) {
+         return;
+      }
+
+      for (const auto& idx : secondary_indices) {
+         auto sec_tbl_key = table_key(make_prefix(table_name, idx->index_name), idx->get_key(value));
+         internal_use_do_not_use::kv_erase(db_name, contract_name.value, sec_tbl_key.data(), sec_tbl_key.size());
+      }
+
+      internal_use_do_not_use::kv_erase(db_name, contract_name.value, tbl_key.data(), tbl_key.size());
+   }
+
+protected:
+   kv_table() = default;
+
+   template <typename... Indices>
+   void setup_indices(uint64_t index_name, bool is_named, null_index* index, Indices... indices) {
+      ++index_name;
+      if constexpr (sizeof...(indices) > 0) {
+         setup_indices(index_name, is_named, indices...);
+      }
+   }
+
+   template <typename I, typename... Indices>
+   void setup_indices(uint64_t index_name, bool is_named, I index, Indices... indices) {
+      if (is_named) {
+         eosio::check(index->index_name.value > 0, "All indices must be named if one is named.");
+      } else {
+         eosio::check(index->index_name.value <= 0, "All indices must be named if one is named.");
```

Comment:
This looks a bit odd, since name.value is unsigned and therefore cannot be less than 0.